accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/authorizableinstaller/impl/NotRegularMembers.java

```diff
@@ -0,0 +1,103 @@
+package biz.netcentric.cq.tools.actool.authorizableinstaller.impl;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import javax.jcr.RepositoryException;
+
+import org.apache.jackrabbit.api.security.user.Authorizable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import biz.netcentric.cq.tools.actool.helper.Constants;
+
+/**
+ * Auxiliar class to store a list of non regular members.
+ */
+class NotRegularMembers {
+
+    private static final Logger LOG = LoggerFactory.getLogger(NotRegularMembers.class);
+
+    private final List<String> members = new CopyOnWriteArrayList<>();
+    private final ExecutorService executor = Executors.newFixedThreadPool(10);
+
+    private RepositoryException thrownException;
+    private final Object thrownExceptionLocker = new Object();
+
+    /**
+     * Add the given member to the set if it is considered a non regular user, discard otherwise
+     *
+     * @param member candidate to be added
+     */
+    public void addOrDiscard(final Authorizable member) {
+        if (executor.isShutdown()) {
+            throw new IllegalStateException("Set is closed and new member cannot be added");
+        }
+        final MemberCheckerWorker memberCheckerWorker = new MemberCheckerWorker(member);
+        executor.execute(memberCheckerWorker);
+    }
+
+    /**
+     * Get the resultant set of filtered users
+     *
+     * @return the set of id of the users
+     * @throws RepositoryException in case it could not verify the id of some user
+     */
+    public Set<String> getSet() throws RepositoryException {
+        waitAllMemberCheckersToFinish();
+        throwExceptionIfAnyWorkerFailed();
+        return new HashSet<>(members);
+    }
+
+    private void throwExceptionIfAnyWorkerFailed() throws RepositoryException {
+        if (thrownException != null) {
+            throw thrownException;
+        }
+    }
+
+    private void waitAllMemberCheckersToFinish() {
+        executor.shutdown();
+        while (!executor.isTerminated()) {
+            try {
+                executor.awaitTermination(10, TimeUnit.SECONDS);
+            } catch (final InterruptedException e) {
+                LOG.error("Thread executor await termination was interrupted.", e);
+            }
+        }
+    }
+
+    private class MemberCheckerWorker implements Runnable {
+        private final Authorizable member;
+
+        MemberCheckerWorker(final Authorizable member) {
+            this.member = Objects.requireNonNull(member);
+        }
+
+        @Override
+        public void run() {
+            try {
+                if (member != null && !isRegularUser(member)) {
+                    members.add(member.getID());
```

Comment:
is this really the bottleneck?

Reply:
I think @tomasalmeida  did a performance test and this supposedly helps, but I also don't understand why (I would have thought `installedGroup.getDeclaredMembers()` is the bottleneck). Even if the query in getDeclaredMembers() was lazy, `currentMemberInRepo.hasNext()` is also called in one thread only. @tomasalmeida do you have more insight what exactly is slow? (is it `member.getPath()` in `isRegularUser` below?)

Reply:
let us enable this only via a feature toggle and not by default. Then we can test some more configurations with this option before rolling it out to everyone.

Reply:
Sure thing! I will add a sling feature flag for that, and perform more tests :-)

Reply:
A simple flag in the YAML config is imho best in this case.

Reply:
we tested more scenarios yesterday and `installedGroup.getDeclaredMembers()` is indeed a bottleneck. It takes 8ms to get each member multiplied by our 100k users (which derives to 13 minutes). We are testing if instead of using this method, we can loop using the group property, but we are not certain of the result.